### PR TITLE
REF: strictness and checks for Timedelta _simple_new

### LINKED ISSUE #9462: Numpy dtype weirdness in TimedeltaIndex._simple_new
I was poking around in tdi.py with pdb and I noticed this code at [line 255 of current master](https://github.com/pydata/pandas/blob/0efd4b3c8d2b0308441a6960327a2447bd419a1b/pandas/tseries/tdi.py#L255):

``` python
    @classmethod
    def _simple_new(cls, values, name=None, freq=None, **kwargs):
-->     if not getattr(values,'dtype',None):   # This doesn't do what it looks like it does
            values = np.array(values,copy=False)
        if values.dtype == np.object_:
            values = tslib.array_to_timedelta64(values)
        if values.dtype != _TD_DTYPE:
            values = com._ensure_int64(values).view(_TD_DTYPE)

        result = object.__new__(cls)
        result._data = values
        result.name = name
        result.freq = freq
        result._reset_identity()
        return result
```

This looks to me like it's testing whether `values` is already an `ndarray`, and if it isn't, it's taking a view into its values.  But that's not what it's doing at all.  If `values` is a `Series`, it has a dtype, so the view-making code should be skipped!  Except that it isn't:

``` python
In [1]: import numpy as np

In [2]: map(np.dtype, [None, 'i8', 'i4,i4', ('m8[ns]', (2, 2)), ('S', 4),
   ...:                [('name', 'u1', (2, 2))],
   ...:                {'names': ['name'], 'formats': ['f8']},
   ...:                ('i4,i4', (2, 2))])
Out[2]: 
[dtype('float64'),                                # note that dtype(None) is float!
 dtype('int64'),                                  # fixed dtype
 dtype([('f0', '<i4'), ('f1', '<i4')]),           # record dtype
 dtype(('<m8[ns]', (2, 2))),                      # array dtype
 dtype('S4'),                                     # flexible dtype
 dtype([('name', 'u1', (2, 2))]),                 # record dtype with array fields
 dtype([('name', '<f8')]),                        # another record dtype
 dtype(([('f0', '<i4'), ('f1', '<i4')], (2, 2)))] # array dtype with record fields

In [3]: map(bool, _)
Out[3]: [False, False, True, False, False, True, True, False]   # madness
```

My tentative guess is that if the outermost layer of the dtype is a record type, then it is `True`y, and otherwise it's `False`y.  This behavior seems too pathological to depend on.

Unfortunately, that's exactly what's happening when constructing a `TimedeltaIndex` from a `Series` of `Timedelta`s:

``` python
In [1]: import pandas as pd

In [2]: import pdb

In [3]: pd.Series([pd.Timedelta(1, 'h')])
Out[3]: 
0   01:00:00
dtype: timedelta64[ns]

In [4]: pdb.run('pd.Index(_)')
> <string>(1)<module>()
(Pdb) b pd.TimedeltaIndex._simple_new
Breakpoint 1 at /Users/afni/ActivityMonitorProcessing/sojourns/source/lib/python2.7/site-packages/pandas/tseries/tdi.py:253
(Pdb) c
> /Users/afni/ActivityMonitorProcessing/sojourns/source/lib/python2.7/site-packages/pandas/tseries/tdi.py(255)_simple_new()
-> if not getattr(values,'dtype',None):
(Pdb) p values
0   01:00:00
dtype: timedelta64[ns]   # it's still a Series
(Pdb) n
> /Users/afni/ActivityMonitorProcessing/sojourns/source/lib/python2.7/site-packages/pandas/tseries/tdi.py(256)_simple_new()
-> values = np.array(values,copy=False)
(Pdb) n
> /Users/afni/ActivityMonitorProcessing/sojourns/source/lib/python2.7/site-packages/pandas/tseries/tdi.py(257)_simple_new()
-> if values.dtype == np.object_:
(Pdb) p values
array([3600000000000], dtype='timedelta64[ns]')   # now it's an ndarray
```

Everything seems to _work_, but what's actually going on is so different from what I would expect to be going on that it seems like a time bomb.

### PR DESCRIPTION
**Pull request by jbrockmendel**
Broken off from #23426.

- [x] closes #9462
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/timedeltas.py [hunk 1]
```diff
@@ -180,12 +194,23 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
...
-        result = super(TimedeltaIndex, cls)._simple_new(values, freq, **kwargs)
+    def _simple_new(cls, values, name=None, freq=None, dtype=_TD_DTYPE):
+        # `dtype` is passed by _shallow_copy in corner cases, should always
+        #  be timedelta64[ns] if present
+        assert dtype == _TD_DTYPE
+
+        assert isinstance(values, np.ndarray), type(values)
+        if values.dtype == 'i8':
+            values = values.view('m8[ns]')
+        assert values.dtype == 'm8[ns]', values.dtype
```
Comment (gfyoung) [inline pandas/core/indexes/timedeltas.py]: Same questions for all your assert statements:

* Are these internal? If not, it would be nice to have user-friendly error messages.
* Can these assert statements be tested in any way?

Comment (gfyoung) [reply pandas/core/indexes/timedeltas.py]: The question of testing is actually more general to all of these changes.  Even though it's been labeled as internal, not sure if any of it surfaces in any way.

**Comment by jbrockmendel** [comment]
The edits in simplenew are thoroughly internal; users shouldn’t get near it.

Not sure what testing these assertions would look like.  Can you elaborate what you have in mind?

File: pandas/core/indexes/timedeltas.py (hunk 1 above)
Comment (gfyoung) [reply pandas/core/indexes/timedeltas.py]: > The edits in simplenew are thoroughly internal; users shouldn’t get near it.
>
> Not sure what testing these assertions would look like. Can you elaborate what you have in m > mind?

(moving your responses to the conversation bubble in the UI, organizational thing)

@jbrockmendel : What I was wondering was whether we could trigger these `assert` statements (e.g. an invalid input to a publicly facing function or method)?

Might be tricky if these edits are purely internal, and if it is too difficult, not a big deal.  Just out of curiosity since tests are good if we can have them.

Comment (jbrockmendel) [reply pandas/core/indexes/timedeltas.py]: I see.  Yah, we could
Write tests where we directly pass invalid inputs to simplenew. (Typing with thumbs, feel free to reformat if necessary)

File: pandas/core/indexes/timedeltas.py [hunk 2]
```diff
@@ -166,6 +167,19 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
         elif copy:
             data = np.array(data, copy=True)
 
+        data = np.array(data, copy=False)
```
Comment (jreback) [inline pandas/core/indexes/timedeltas.py]: do you need this here? the array conversion

File: pandas/core/indexes/timedeltas.py [hunk 3]
```diff
@@ -180,12 +194,23 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
...
+        if values.dtype == 'i8':
+            values = values.view('m8[ns]')
+        assert values.dtype == 'm8[ns]', values.dtype
+
+        result = super(TimedeltaIndex, cls)._simple_new(values, freq)
         result.name = name
         result._reset_identity()
         return result
 
+    _shallow_copy = Index._shallow_copy
```
Comment (jreback) [inline pandas/core/indexes/timedeltas.py]: why is this needed?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/timedeltas.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/indexes/timedeltas.py]: Because until #23430, without pinning this it would use DatetimeLikeArrayMixin._shallow_copy

File: pandas/core/indexes/timedeltas.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/timedeltas.py]: This whole block is taken out of the top of `TimedeltaArray._simple_new`.  I think we could get here with e.g. another TimedeltaIndex for `data`.

File: pandas/core/arrays/timedeltas.py [hunk 4]
```diff
@@ -111,16 +111,16 @@ def dtype(self):
...
-        values = np.array(values, copy=False)
-        if values.dtype == np.object_:
-            values = array_to_timedelta64(values)
-        if values.dtype != _TD_DTYPE:
-            if is_timedelta64_dtype(values):
-                # non-nano unit
-                values = values.astype(_TD_DTYPE)
-            else:
-                values = ensure_int64(values).view(_TD_DTYPE)
+    def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py]: in future, or here ok, can you add a doc-string

File: pandas/core/indexes/timedeltas.py [hunk 5]
```diff
@@ -166,6 +167,19 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
         elif copy:
             data = np.array(data, copy=True)
 
+        data = np.array(data, copy=False)
+        if data.dtype == np.object_:
+            data = array_to_timedelta64(data)
```
Comment (jreback) [inline pandas/core/indexes/timedeltas.py]: why are these checks NOT done in _simple_new? this is inconsistent with other code.

We should be really really clear on what is acceptable in _simple_new vs. what is not. IIRC from another of your PR's you did checks on object type in _simple_new for example.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/timedeltas.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: will do in next pass

File: pandas/core/indexes/timedeltas.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/indexes/timedeltas.py]: AFAICT the current verbose-checking is largely driven by the weird cases (that these PRs get rid of) where `None` or `[]` is passed to `_shallow_copy`.

This and the associated DatetimeIndex PR impose a simple/strict API for _simple_new: it expects an np.ndarray that may be either i8 or M8[ns]/m8[ns].

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/timedeltas.py [hunk 6]
```diff
@@ -131,6 +131,10 @@ def __new__(cls, values, freq=None):
 
         freq, freq_infer = dtl.maybe_infer_freq(freq)
 
+        values = np.array(values, copy=False)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: ATM we are checking for it in `TimedeltaArray._simple_new`, so de-facto accepting it in `TimedeltaArray.__new__`.  This is leaving the effective `__new__` policy unchanged while clearing up the `_simple_new` policy

**Comment by jbrockmendel** [comment]
Gentle ping,  I’m hoping to make a big push on datetime/timedelta array this weekend, getting the constructors locked down is the first step.

File: pandas/core/arrays/timedeltas.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/timedeltas.py]: ok, assume this is on the list to de-duplicate

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
ok merging.
